# Data Cleaning
- Purpose: Clean and prepare Telco data for Youtube Analysis
- Process: Handle missing values, type conversion, outlier detection

In [78]:
import pandas as pd
import numpy as np

# Load Telco data
df = pd.read_csv('../data/raw/telco_data.csv')
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No     1.0           No   
1  5575-GNVDE    Male              0      No         No    34.0          Yes   
2  3668-QPYBK    Male              0      No         No     2.0          Yes   
3  7795-CFOCW    Male              0      No         No    45.0           No   
4  9237-HQITU  Female              0      No         No     2.0          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [79]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            6293 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           6043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            4543 non-null   float64
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   6043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       5543 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

In [80]:
df.isnull().sum()

customerID             0
gender               750
SeniorCitizen          0
Partner             1000
Dependents             0
tenure              2500
PhoneService           0
MultipleLines          0
InternetService     1000
OnlineSecurity         0
OnlineBackup           0
DeviceProtection       0
TechSupport            0
StreamingTV         1500
StreamingMovies        0
Contract               0
PaperlessBilling       0
PaymentMethod          0
MonthlyCharges      1500
TotalCharges           0
Churn                  0
dtype: int64

In [81]:
df.describe()

SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  4543.000000     5543.000000
mean        0.162147    32.546555       64.876403
std         0.368612    24.505519       30.101331
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.750000
50%         0.000000    29.000000       70.550000
75%         0.000000    56.000000       89.925000
max         1.000000    72.000000      118.600000

### Missing Values + Type
- gender: 6293 drop or fill out
- tenure: 4543 replace with median
- MonthlyCharges, StreamingTV: 5543 missing values
- TotalCharges: 
              1. str--> doubted numeric values might be changed as string
              2. Why missing values is "0"?? might be missing values transit into NaN due to the 1.
              3. Need to convert the values into float type and check the missing values again

In [82]:
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')

In [83]:
df['TotalCharges'].isnull().sum()

np.int64(11)

In [84]:
print(df['TotalCharges'].dtype)
print(df.info())

float64
<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            6293 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           6043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            4543 non-null   float64
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   6043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       5543 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17

In [85]:
df.describe()

SeniorCitizen       tenure  MonthlyCharges  TotalCharges
count    7043.000000  4543.000000     5543.000000   7032.000000
mean        0.162147    32.546555       64.876403   2283.300441
std         0.368612    24.505519       30.101331   2266.771362
min         0.000000     0.000000       18.250000     18.800000
25%         0.000000     9.000000       35.750000    401.450000
50%         0.000000    29.000000       70.550000   1397.475000
75%         0.000000    56.000000       89.925000   3794.737500
max         1.000000    72.000000      118.600000   8684.800000

In [86]:
# tenure missing values
df[df['tenure'].isnull()].index

Index([   8,   12,   14,   15,   17,   19,   23,   26,   29,   30,
       ...
       7010, 7013, 7018, 7023, 7025, 7031, 7034, 7036, 7037, 7041],
      dtype='int64', length=2500)

In [87]:
df[df['tenure'].isnull()].head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
8   7892-POOKP  Female              0     Yes         No     NaN          Yes   
12  8091-TTVAX    Male              0     Yes         No     NaN          Yes   
14  5129-JLPIS    Male              0      No         No     NaN          Yes   
15  3655-SNQYZ  Female              0     Yes        Yes     NaN          Yes   
17  9959-WOFKT     NaN              0     NaN        Yes     NaN          Yes   

   MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
8            Yes     Fiber optic             No  ...              Yes   
12           Yes     Fiber optic             No  ...              Yes   
14            No     Fiber optic            Yes  ...              Yes   
15           Yes     Fiber optic            Yes  ...              Yes   
17           Yes             NaN            Yes  ...              Yes   

   TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
8          Yes         NaN             Yes  Month-to-month              Yes   
12          No         Yes             Yes        One year               No   
14         Yes         NaN             Yes  Month-to-month              Yes   
15         Yes         NaN             Yes        Two year               No   
17          No         NaN             Yes        Two year               No   

                PaymentMethod MonthlyCharges  TotalCharges  Churn  
8            Electronic check            NaN       3046.05    Yes  
12    Credit card (automatic)         100.35       5681.10     No  
14           Electronic check            NaN       2686.05     No  
15    Credit card (automatic)            NaN       7895.15     No  
17  Bank transfer (automatic)            NaN       7382.25     No  

[5 rows x 21 columns]

In [88]:
df[df['tenure'].isnull() & df['TotalCharges'].isnull()].shape[0]

3

In [89]:
df[df['tenure'].isnull() & df['MonthlyCharges'].isnull()].shape[0]

1500

In [90]:
df[df['tenure'].isnull() & df['MonthlyCharges'].notnull()].shape[0]

1000

In [91]:
df[df['tenure'].isnull() & df['TotalCharges'].notnull()].shape[0]

2497

In [92]:
# Step 1. TotalCharges & MonthlyCharges IS not Null & tenure is null
# tenure = TotalCharges / MonthlyCharges

temp = (df['TotalCharges'].notnull()) & \
       (df['MonthlyCharges'].notnull()) & \
       (df['tenure'].isnull())                    

# count temp
print(f"Count: {temp.sum()} lines")

# get tenure
df.loc[temp, 'tenure'] = (df.loc[temp, 'TotalCharges'] / df.loc[temp, 'MonthlyCharges']).round(0)

print(df[temp][['TotalCharges', 'MonthlyCharges', 'tenure']].head())
print(f"count of missing values in tenure: {df['tenure'].isnull().sum()}")


Count: 999 lines
    TotalCharges  MonthlyCharges  tenure
12       5681.10          100.35    57.0
29       1093.10           64.70    17.0
30       6766.95           96.35    70.0
43       4861.45           79.85    61.0
47        144.15           80.65     2.0
count of missing values in tenure: 1501


In [ ]:
# Step 2. TotalCharges & tenure IS not Null & MonthlyCharges is null
# MonthlyCharges = TotalCharges / tenure

temp2 = (df['TotalCharges'].notnull()) & \
       (df['tenure'].notnull()) & \
       (df['MonthlyCharges'].isnull())                    

# count temp
print(f"Count: {temp2.sum()} lines")

# get tenure
df.loc[temp2, 'MonthlyCharges'] = df.loc[temp2, 'TotalCharges'] / df.loc[temp2, 'tenure']

print(df[temp][['TotalCharges', 'MonthlyCharges', 'tenure']].head())
print(f"count of missing values in MonthlyCharges: {df['MonthlyCharges'].isnull().sum()}")


Count: 0 lines
    TotalCharges  MonthlyCharges  tenure
12       5681.10          100.35    57.0
29       1093.10           64.70    17.0
30       6766.95           96.35    70.0
43       4861.45           79.85    61.0
47        144.15           80.65     2.0
count of missing values in MonthlyCharges: 1500


In [94]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            6293 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           6043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            5542 non-null   float64
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   6043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       5543 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

In [ ]:
# Monthly missing values --> fill out with Median

monthly_median = df['MonthlyCharges'].median()
print(f"Monthly Charages median: {monthly_median}")

df['MonthlyCharges'] = df['MonthlyCharges'].fillna(monthly_median)

print(f"Current missing Values in MonthlyCharges: {df['MonthlyCharges'].isnull().sum()}")
print(df[['tenure', 'MonthlyCharges', 'TotalCharges']].isnull().sum())

Monthly Charages median: 70.55
Current missing Values in MonthlyCharges: 0
tenure            1501
MonthlyCharges       0
TotalCharges        11
dtype: int64


In [100]:
# Now fill out the rest of tenure 

temp_tenure_final = (df['TotalCharges'].notnull()) & \
                    (df['MonthlyCharges'].notnull()) & \
                    (df['tenure'].isnull())      

# count temp
print(f"Count: {temp_tenure_final.sum()} lines")

# get tenure
df.loc[temp_tenure_final, 'tenure'] = (df.loc[temp_tenure_final, 'TotalCharges'] / df.loc[temp_tenure_final, 'MonthlyCharges']).round(0)

print(df[temp_tenure_final][['TotalCharges', 'MonthlyCharges', 'tenure']].head())
print(f"count of missing values in tenure: {df['tenure'].isnull().sum()}") 

print(df.info())

Count: 1498 lines
    TotalCharges  MonthlyCharges  tenure
8        3046.05           70.55    43.0
14       2686.05           70.55    38.0
15       7895.15           70.55   112.0
17       7382.25           70.55   105.0
19       1862.90           70.55    26.0
count of missing values in tenure: 3
<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            6293 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           6043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7040 non-null   float64
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   6043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   

In [101]:
# Now drop all rows with missing values
df = df.dropna()

df.info()

print(f"Total rows: {len(df)}")
print(f"Final total count of missing values: {df.isnull().sum().sum()}  ")

<class 'pandas.DataFrame'>
Index: 5534 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        5534 non-null   str    
 1   gender            5534 non-null   str    
 2   SeniorCitizen     5534 non-null   int64  
 3   Partner           5534 non-null   str    
 4   Dependents        5534 non-null   str    
 5   tenure            5534 non-null   float64
 6   PhoneService      5534 non-null   str    
 7   MultipleLines     5534 non-null   str    
 8   InternetService   5534 non-null   str    
 9   OnlineSecurity    5534 non-null   str    
 10  OnlineBackup      5534 non-null   str    
 11  DeviceProtection  5534 non-null   str    
 12  TechSupport       5534 non-null   str    
 13  StreamingTV       5534 non-null   str    
 14  StreamingMovies   5534 non-null   str    
 15  Contract          5534 non-null   str    
 16  PaperlessBilling  5534 non-null   str    
 17  PaymentMeth

In [102]:
# Save cleaned data
df.to_csv('../data/raw/cleaned_data.csv', index=False)

# Verify saved file
print("File saved successfully!")
print(f"Rows: {len(df)}, Columns: {len(df)}")

File saved successfully!
Rows: 5534, Columns: 5534
